### PR TITLE
Avoid reading JDBC connector table properties

### PR DESCRIPTION
**Pull request by kokosing**
Avoid reading JDBC connector table properties

This is achievied by Implementing
getTableBasicMetadata for JDBC connectors

### CODE REVIEW

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java [hunk 1]
```diff
@@ -330,20 +331,45 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
...
 
+    @Override
+    public ConnectorTableBasicMetadata getTableBasicMetadata(ConnectorSession session, ConnectorTableHandle table)
+    {
+        JdbcTableHandle handle = (JdbcTableHandle) table;
+
+        return new ConnectorTableBasicMetadata(
+                getSchemaTableName(handle),
+                getColumnMetadata(session, handle).stream()
+                        .map(ColumnMetadata::toBasicMetadata)
```
Comment (ssheikin) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java:342]: Seems you can first `getColumnMetadata` and later  `.map(ColumnMetadata::toBasicMetadata)` only because `ColumnMetadata` does not have column properties. So actually it masks current inexistent problem with loading column properties for jdbc connectors, but problem would still exist if jdbc connector would have properties.

File: plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixMetadata.java [hunk 2]
```diff
@@ -83,19 +84,32 @@ public JdbcTableHandle getTableHandle(ConnectorSession session, SchemaTableName
...
     {
         JdbcTableHandle handle = (JdbcTableHandle) table;
-        List<ColumnMetadata> columnMetadata = phoenixClient.getColumns(session, handle).stream()
+        return new ConnectorTableMetadata(
+                getSchemaTableName(handle),
+                getColumnMetadata(session, false, handle),
+                phoenixClient.getTableProperties(session, handle));
+    }
+
+    private List<ColumnMetadata> getColumnMetadata(ConnectorSession session, boolean rowkeyRequired, JdbcTableHandle handle)
```
Comment (ssheikin) [inline plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixMetadata.java:107]: order of parameters differs from https://github.com/trinodb/trino/pull/6975/commits/eaf66717f28023799cdc2552b882587ba0cb6eb5#diff-b2770586e716078fd17f177ce4cd0868628f704bc0cd81f3fb5f3d4def54e882L91

`rowkeyRequired` is always `false`.

**Comment by kokosing** [comment]
This approach 4 from https://docs.google.com/document/d/1LqzlflWo4m6DySKihE1LSsDhqhOisAOET45pERJKEa4/edit#

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java (hunk 1 above)
Comment (kokosing) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java:342]: That is true. If we ever add column properties to JDBC connectors I don't believe that `JdbcColumnHandle` will be able to create `ColumnMetadata`. And surely we don't want `io.trino.plugin.jdbc.JdbcClient#getColumns` to query remote data source for column properties.

**Comment by kokosing** [comment]
AC

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 3]
```diff
@@ -507,12 +507,7 @@ protected Scope visitRefreshMaterializedView(RefreshMaterializedView refreshMate
                 throw semanticException(TABLE_NOT_FOUND, refreshMaterializedView, "Table '%s' does not exist", targetTable);
             }
 
-            if (targetTableHandle.isPresent() && metadata.getMaterializedViewFreshness(session, name).isMaterializedViewFresh()) {
-                analysis.setSkipMaterializedViewRefresh(true);
-            }
-            else {
-                analysis.setSkipMaterializedViewRefresh(false);
-            }
+            analysis.setSkipMaterializedViewRefresh(metadata.getMaterializedViewFreshness(session, name).isMaterializedViewFresh());
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:512]: Why `targetTableHandle.isPresent()` is being removed?

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 4]
```diff
@@ -2266,7 +2261,7 @@ private void analyzeWindowFunctions(QuerySpecification node, List<Expression> ou
                 ResolvedWindow window = analysis.getWindow(windowFunction);
                 // TODO get function requirements from window function metadata when we have it
                 String name = windowFunction.getName().toString().toLowerCase(ENGLISH);
-                if (name.equals("lag") || name.equals("lead")) {
+                if ("lag".equals(name) || "lead".equals(name)) {
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:2266]: We don't have convention either way, and the `name` cannot be null, so this change isn't necessary

**Review by findepi — COMMENTED** [review]
"Cleanup in StatementAnalyzer"

File: core/trino-main/src/main/java/io/trino/metadata/Metadata.java [hunk 5]
```diff
@@ -113,6 +113,14 @@
 
     Optional<Object> getInfo(Session session, TableHandle handle);
 
+    /**
+     * Return basic metadata for the specified table handle.
+     * This is method should be preferred over {@link #getTableMetadata(Session, TableHandle)} as it is cheaper.
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/Metadata.java:118]: s/is//

File: core/trino-main/src/main/java/io/trino/metadata/TableBasicMetadata.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    public List<ColumnBasicMetadata> getColumns()
+    {
+        return metadata.getColumns();
+    }
+
+    public ColumnBasicMetadata getColumn(String name)
+    {
+        return metadata.getColumns().stream()
+                .filter(columnMetadata -> columnMetadata.getName().equals(name))
+                .findFirst()
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/TableBasicMetadata.java:68]: `.collect(toOptional())` would throw on duplicates, which is more appropriate here.

File: core/trino-spi/src/main/java/io/trino/spi/connector/ColumnMetadata.java [hunk 7]
```diff
@@ -135,6 +135,11 @@ public boolean isHidden()
         return properties;
     }
 
+    public ColumnBasicMetadata toBasicMetadata()
+    {
+        return new ColumnBasicMetadata(name, type, nullable, hidden);
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ColumnMetadata.java:140]: If the intention is that `ColumnBasicMetadata` is a subset of information, maybe `ColumnBasicMetadata` should be accepted in the `ColumnMetadata` ctor and be stored as a field? This could allow connectors to maintain mapping once.

However, I am not sure whether we can want to lock ourselves in one being subset of the other -- logically it make sense, but there can be a situation, where this isn't helpful. For this however, we would want this method (`toBasicMetadata()`) not to exist at all. (or exist as deprecated for the transition period, if needed)

thoughts?

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorTableMetadata.java [hunk 8]
```diff
@@ -71,6 +72,15 @@ public SchemaTableName getTable()
         return comment;
     }
 
+    public ConnectorTableBasicMetadata toBasicMetadata()
+    {
+        return new ConnectorTableBasicMetadata(
+                table,
+                columns.stream()
+                        .map(ColumnMetadata::toBasicMetadata)
+                        .collect(Collectors.toUnmodifiableList()));
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorTableMetadata.java:81]: static import

File: core/trino-spi/src/main/java/io/trino/spi/connector/ColumnBasicMetadata.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        ColumnBasicMetadata that = (ColumnBasicMetadata) o;
+        return nullable == that.nullable && hidden == that.hidden && name.equals(that.name) && type.equals(that.type);
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ColumnBasicMetadata.java:72]: put on separate lines

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java (hunk 3 above)
Comment (kokosing) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:512]: It is already checked two lines above

File: core/trino-spi/src/main/java/io/trino/spi/connector/ColumnBasicMetadata.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+    @Override
+    public String toString()
+    {
+        StringBuilder sb = new StringBuilder("ColumnBasicMetadata{");
+        sb.append("name='").append(name).append('\'');
+        sb.append(", type=").append(type);
+        sb.append(", nullable=").append(nullable);
+        sb.append(", hidden=").append(hidden);
+        sb.append('}');
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ColumnBasicMetadata.java:89]: My IDE generates toString using Joiner.

**Review by findepi — COMMENTED** [review]
"Introduce ConnectorTableBasicMetadata"

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcColumnHandle.java [hunk 11]
```diff
@@ -108,6 +109,11 @@ public ColumnMetadata getColumnMetadata()
                 .build();
     }
 
+    public ColumnBasicMetadata getColumnBasicMetadata()
+    {
+        return new ColumnBasicMetadata(columnName, columnType, nullable, false);
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcColumnHandle.java:114]: why not `getColumnMetadata(). toBasic()`?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java [hunk 12]
```diff
@@ -330,20 +331,37 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
...
-        SchemaTableName schemaTableName = handle.isNamedRelation()
+        return new ConnectorTableMetadata(
+                getSchemaTableName(handle),
+                jdbcClient.getColumns(session, handle).stream()
+                        .map(JdbcColumnHandle::getColumnMetadata)
+                        .collect(toImmutableList()),
+                jdbcClient.getTableProperties(session, handle));
+    }
+
+    protected static SchemaTableName getSchemaTableName(JdbcTableHandle handle)
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java:359]: This method is extracted from a hack (see TODO inside) and hopefully can stay `private`, unless you already see a need to expose this.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcTableProperties.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+import org.testng.annotations.Test;
+
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static io.trino.plugin.jdbc.H2QueryRunner.createH2QueryRunner;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.testng.Assert.fail;
+
+@Test(singleThreaded = true)
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcTableProperties.java:32]: let's document why
```
// Single-threaded because of shared mutable state, e.g. onGetTableProperties
```

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcTableProperties.java [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        onGetTableProperties = () -> { fail("Unexpected call of: getTableProperties"); };
+        assertUpdate("CREATE TABLE copy_of_nation AS SELECT * FROM nation", 25);
+        assertQuerySucceeds("SELECT * FROM copy_of_nation");
+        assertQuerySucceeds("SELECT nationkey FROM copy_of_nation");
+    }
+
+    @Test
+    public void testGetTablePropertiesIsCalled()
+    {
+        AtomicBoolean getTablePropertiesIsCalled = new AtomicBoolean();
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcTableProperties.java:73]: AtomicInteger to assess call count

File: plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixMetadata.java [hunk 15]
```diff
@@ -83,19 +84,32 @@ public JdbcTableHandle getTableHandle(ConnectorSession session, SchemaTableName
...
         JdbcTableHandle handle = (JdbcTableHandle) table;
-        List<ColumnMetadata> columnMetadata = phoenixClient.getColumns(session, handle).stream()
-                .filter(column -> rowkeyRequired || !ROWKEY.equalsIgnoreCase(column.getColumnName()))
+        return new ConnectorTableMetadata(
+                getSchemaTableName(handle),
+                getColumnMetadata(session, handle),
+                phoenixClient.getTableProperties(session, handle));
+    }
+
+    private List<ColumnMetadata> getColumnMetadata(ConnectorSession session, JdbcTableHandle handle)
```
Comment (findepi) [inline plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixMetadata.java:107]: nit: or `Stream`

**Review by findepi — APPROVED** [review]
Alternative would be to move properties out of the current metadata objects, as attempted here https://github.com/trinodb/trino/pull/6966.
The sheer volume of changes can put off, but it's also partially already done.
However, it also means that change cannot be done in a backwards compatible manner, which is an obvious downside, and doesn't guarantee no similar problem in the future, where again CREATE TABLE needs a different set of information about a table than regular SELECT (e.g. indexes; from engine perspective).
For this reason, I  LGTM this PR, even though it adds duplication of table/column metadata classes.

**Comment by findepi** [comment]
@martint @losipiuk @electrum you may be interested in taking a look

**Review by martint — CHANGES_REQUESTED** [review]
The concept of basic vs non-basic metadata is confusing. What goes where? When does something qualify as a basic vs non-basic piece of metadata?

I would rather we split the API into two parts as in #6966.

For the purpose of gradual deprecation, we can add the new API along side the existing one, mark the "properties" part as deprecated for a while (the engine would call one, then the other, if the first one doesn't produce any results), migrate all connectors to the new API and leave some time buffer for third-party connectors to do so. Then we can get rid of the old API.

**Comment by kokosing** [comment]
> I would rather we split the API into two parts as in #6966.

@martint I admit that #6966 sounds promising. But here is my concern:

Currently are two ways engine communicates with connector in regards to a table. By using table handle or using table metadata. These two objects match to each other, they are some kind equivalent to each other. We use handle for existing tables and we use metadata for not yet existing tables (create table, CTAS). We also use metadata for existing tables too, to simplify code paths so same code path is executed for CTAS and insert.

If table does not exists, we are forced to use table metadata. If we split it with #6966 then there will be no single class that would be a complete representation of a table, hence in all places where today metadata is used we would need to pass the pair.

The worst part is that someday someone would like to name that pair. Then the name could be `ConnectorTableFullMetadata`. Then we could end up here, but in a bit different shape.

> The concept of basic vs non-basic metadata is confusing. What goes where? When does something qualify as a basic vs non-basic piece of metadata?

I hope this is a naming issue. Rule of thumb, all goes to non-basic and only things that are needed to run non DDL queries goes to the basic.

File: core/trino-spi/src/main/java/io/trino/spi/connector/ColumnMetadata.java (hunk 7 above)
Comment (kokosing) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ColumnMetadata.java:140]: > If the intention is that ColumnBasicMetadata is a subset of information, maybe ColumnBasicMetadata should be accepted in the ColumnMetadata ctor and be stored as a field?

Sounds good to me. That was my intention. I wanted to leave it optional for connectors to provide basic metadata. For example in case of Hive it does not really matter if we provide basic or non-basic as it translates to same number of HMS calls.

**Comment by findepi** [comment]
> The worst part is that someday someone would like to name that pair. Then the name could be ConnectorTableFullMetadata. Then we could end up here, but in a bit different shape.


Passing a pair as individual parameters -- ConnectorTableMetadata and table/column properties -- solves the problem we have today, but does not provide for future extendibility easily
The thing to consider is -- what if we introduce new type of information describing the table, which we cannot or do not want to put into our slimmed down ConnectorTableMetadata.

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java [hunk 16]
```diff
@@ -161,6 +161,18 @@ default ConnectorTableHandle makeCompatiblePartitioning(ConnectorSession session
...
     }
 
+    /**
+     * Return basic table metadata for the specified table handle.
+     * This method is useful when getting full table metadata is expensive.
+     * Engine will try to leverage basic table metadata whenever possible.
+     *
+     * @throws RuntimeException if table handle is no longer valid
+     */
+    default ConnectorTableBasicMetadata getTableBasicMetadata(ConnectorSession session, ConnectorTableHandle table)
```
Comment (jirassimok) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:171]: How about calling it `getBasicTableMetadata`? No matter how I turn `getTableBasicMetadata` around in my head, I can't make it parse as anything like "get basic metadata for table."

The more direct `getBasicMetadataForTable` also seems like a good name to me.

I think even `getConnectorTableBasicMetadata` is better than `getTableBasicMetadata`, but it's very verbose, and `get[ReturnType]` is not a very useful function name (but I think it's better than the current version, which is neither the return type nor phrase with similar meaning).

File: core/trino-spi/src/main/java/io/trino/spi/connector/ColumnBasicMetadata.java (hunk 10 above)
Comment (kokosing) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ColumnBasicMetadata.java:89]: I wanted to be in line with `io.trino.spi.connector.ColumnMetadata#toString`. My IDE allows me to choose how to generate `toString`.

Is there any difference? If so then it would be nice to update other `toString` methods in SPI.

File: core/trino-spi/src/main/java/io/trino/spi/connector/ColumnMetadata.java (hunk 7 above)
Comment (kokosing) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ColumnMetadata.java:140]: > be stored as a field

That was my plan too. Also @martint mentioned that is the way he would prefer. However it is not that simple as there is not only ConnectorTableSchema, but also ColumnSchema. So I would need to somehow split `ColumnMetadata` in `ConnectorTableMetadata`.

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java (hunk 16 above)
Comment (kokosing) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:171]: Renamed it to `ConnectorTableSchema` after discussion with @martint .

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcColumnHandle.java (hunk 11 above)
Comment (kokosing) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcColumnHandle.java:114]: I wanted to avoid intermediate object creation. Do you think, it is worth to create?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java (hunk 12 above)
Comment (kokosing) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java:359]: > unless you already see a need to expose this.

I do, phoenix connector potentially can also have `!handle.isNamedRelation()`, but it was missing a TODO like this there.

**Review by kokosing — COMMENTED** [review]
After a discussion with @martint we decided that this should be named `TableSchema`. We even prefered `TableSchemaDefinition` to avoid confusion with schemas (part of catalog and group of tables`, but `ConnectorTableSchemaDefinition` was too long to me.

`TableSchema` represents a schema of table. See `CREATE TABLE <table name> (<table schema>) WITH <table properties>`. However note that columns can also have properties.

Comments addressed, please take a look.

File: core/trino-main/src/main/java/io/trino/metadata/Metadata.java (hunk 5 above)
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/metadata/Metadata.java]: bump

File: core/trino-main/src/main/java/io/trino/metadata/Metadata.java [hunk 17]
```diff
@@ -114,6 +114,16 @@
 
     Optional<Object> getInfo(Session session, TableHandle handle);
 
+    /**
+     * Return table schema definition for the specified table handle.
+     * Table schema definition is a set of information
+     * required by semantic analyzer to analyze the query.
+     * This is method should be preferred over {@link #getTableMetadata(Session, TableHandle)} as it is cheaper.
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/Metadata.java]: Instead of (or additionally to) having this sentence here, add a plain `@see` pointing to `getTableMetadata`
In `getTableMetadata` add `@see` poitining here and recommending use of this method, as it returns less information

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java [hunk 18]
```diff
@@ -161,6 +161,20 @@ default ConnectorTableHandle makeCompatiblePartitioning(ConnectorSession session
         return Optional.empty();
     }
 
+    /**
+     * Return table schema definition for the specified table handle.
+     * Table schema definition is a set of information
+     * required by semantic analyzer to analyze the query.
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:167]: Do we need to define the content of the TableSchema by referring to the engine internal?

connector implementor may not care what is semantic analyzer and what other analyzers we have internally.

if you want to refer to usecase, maybe it would be better to say that this is the information required when planning a SELECT query?

Note however, that even in such case, we call other methods, like getTableProperties

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java [hunk 19]
```diff
@@ -161,6 +161,20 @@ default ConnectorTableHandle makeCompatiblePartitioning(ConnectorSession session
...
+     * This method is useful when getting full table metadata is expensive.
+     * Engine will try to leverage basic table metadata whenever possible.
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:169]: "basic table metadata" -- this doens't exist

File: core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+
+import io.trino.spi.type.Type;
+
+import java.util.Objects;
+
+import static io.trino.spi.connector.SchemaUtil.checkNotEmpty;
+import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
+
+public class ColumnSchema
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java:24]: We may need a build for this, like we have for `ColumnMetadata`

as we strive for this class to be as small as possible, i don't believe we won't extend it in the future with additional information

for example we have nullable, we could add some other contraint information

File: core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+    {
+        return name;
+    }
+
+    public Type getType()
+    {
+        return type;
+    }
+
+    public boolean isNullable()
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java:52]: seems unused today. Curious about your thinking whether to include it here or not.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcTableProperties.java [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static io.trino.plugin.jdbc.H2QueryRunner.createH2QueryRunner;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.testng.Assert.fail;
+
+@Test(singleThreaded = true)
+// Single-threaded because of shared mutable state, e.g. onGetTableProperties
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcTableProperties.java:33]: comment should be before `@Test(singleThrea..`

**Review by findepi — APPROVED** [review]

File: core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java (hunk 21 above)
Comment (kokosing) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java:52]: Planner might like to use it.

Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java:52]: I agree. Planner may want to use this and other attributes, this is why i suggested adding bulder (https://github.com/trinodb/trino/pull/6975#discussion_r587386504) for future compatibility.

Anyway, we may never need this. This information is already available via `ConnectorTableProperties` -> predicate -> `Domain`  -> contains null?

so let's remove it for now

Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java:52]: if you prefer to keep it, add a comment. otherwise someone may do "Remove unused field" tomorrow, pruning it :)

**Comment by kokosing** [comment]
AC.

@martint Please take a look.

**Comment by kokosing** [comment]
@martint ping

**Comment by kokosing** [comment]
CI hit: https://github.com/trinodb/trino/issues/7468

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 23]
```diff
@@ -483,7 +483,7 @@ protected Scope visitInsert(Insert insert, Optional<Scope> scope)
                     targetTable,
                     Optional.empty(),
                     Optional.of(insertColumns.stream()
-                            .map(insertColumn -> new Column(insertColumn, tableMetadata.getColumn(insertColumn).getType().toString()))
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:484]: This doesn't seem to belong in this commit

File: core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+
+        public Builder setHidden(boolean hidden)
+        {
+            this.hidden = hidden;
+            return this;
+        }
+
+        public ColumnSchema build()
+        {
+            return new ColumnSchema(
```
Comment (martint) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java:137]: These can all go in the same line.

File: core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java [hunk 25]
```diff
@@ -0,0 +1,138 @@
...
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(name, type, hidden);
+    }
+
+    @Override
+    public String toString()
+    {
+        StringBuilder sb = new StringBuilder("ColumnBasicMetadata{");
```
Comment (martint) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java:79]: s/sb/builder/

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorTableSchema.java [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+
+    public List<ColumnSchema> getColumns()
+    {
+        return columns;
+    }
+
+    @Override
+    public String toString()
+    {
+        final StringBuilder sb = new StringBuilder("ConnectorTableSchema{");
```
Comment (martint) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorTableSchema.java:47]: s/sb/builder

Also, remove `final`

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorTableMetadata.java [hunk 27]
```diff
@@ -78,7 +78,7 @@ public ConnectorTableSchema getTableSchema()
                 table,
                 columns.stream()
                         .map(ColumnMetadata::getColumnSchema)
-                        .collect(Collectors.toUnmodifiableList()));
+                        .collect(toUnmodifiableList()));
```
Comment (martint) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorTableMetadata.java:81]: Unrelated change

**Review by martint — COMMENTED** [review]
Some minor comments. I only reviewed the engine parts.

**Comment by kokosing** [comment]
CI hit: https://github.com/trinodb/trino/issues/4984

**Comment by kokosing** [comment]
@martint rebased, and addressed comments.

File: core/trino-main/src/main/java/io/trino/metadata/TableSchema.java [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+    public List<ColumnSchema> getColumns()
+    {
+        return tableSchema.getColumns();
+    }
+
+    public ColumnSchema getColumn(String name)
+    {
+        return tableSchema.getColumns().stream()
+                .filter(columnMetadata -> columnMetadata.getName().equals(name))
+                .collect(toOptional())
```
Comment (Praveen2112) [inline core/trino-main/src/main/java/io/trino/metadata/TableSchema.java:69]: nit : How about `findFirst`. Won't it be a bit easy to read ?

File: core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java [hunk 29]
```diff
@@ -129,10 +129,7 @@ public Builder setHidden(boolean hidden)
 
         public ColumnSchema build()
         {
-            return new ColumnSchema(
-                    name,
-                    type,
-                    hidden);
+            return new ColumnSchema(name, type, hidden);
```
Comment (Praveen2112) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java:137]: This can be applied in the previous commit right ?

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorTableSchema.java [hunk 30]
```diff
@@ -44,10 +44,10 @@ public SchemaTableName getTable()
     @Override
     public String toString()
     {
-        final StringBuilder sb = new StringBuilder("ConnectorTableSchema{");
-        sb.append("table=").append(table);
-        sb.append(", columns=").append(columns);
-        sb.append('}');
-        return sb.toString();
+        return new StringBuilder("ConnectorTableSchema{")
```
Comment (Praveen2112) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorTableSchema.java:47]: Can we move this previous commit ?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcColumnHandle.java [hunk 31]
```diff
@@ -108,6 +109,11 @@ public ColumnMetadata getColumnMetadata()
                 .build();
     }
 
+    public ColumnSchema getColumnSchema()
+    {
+        return ColumnSchema.from(getColumnMetadata());
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcColumnHandle.java:114]: Any reason for not using `Builder` ?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java (hunk 12 above)
Comment (Praveen2112) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java:545]: Maybe as a separate commit ?

**Review by Praveen2112 — APPROVED** [review]
Minor changes. LGTM...

File: core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java (hunk 24 above)
Comment (martint) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java]: This is still not resolved. Did you forget to push your changes?

**Review by martint — APPROVED** [review]

**Comment by ssheikin** [comment]
:tada:
